Improve make checks, fix annoyances

jeffmccune opened the pull request:
Without this patch there are a number of annoyances with the make
checks.  This patch fixes the following:

 1. Newlines not being added to auto-generated documentation.
 2. Auto generated documentation failing if README.md does not exist.
 3. Checks execute against .git/ index
 4. terraform validate executes redundantly

To address the doc generation annoyances, this patch converts
combine_docfiles to fairly simple bash.  The behavior is pure sed, the
rest of the bash is argument and error handling.

For the others, the patch refactors make.sh to use a general find |
xargs pattern with a single function to manage exclusion lists.

check_shebangs is removed because `#! /bin/bash -e` should not be used
because it does not take effect if the script is called via bash
<filename>.  Instead, calling `set -e` is preferred.

## Expected results:

```
make
Running shellcheck
Running flake8
Running go fmt and go vet
Running terraform validate
Running hadolint on Dockerfiles
Checking for required files LICENSE README.md
Testing the validity of the header check
..
----------------------------------------------------------------------
Ran 2 tests in 0.057s

OK
Checking file headers
The following lines have trailing whitespace
Generating markdown docs with terraform-docs
No changes made: ./README.md
No changes made: ./examples/simple_example/README.md
```

jeffmccune committed:
Improve make checks, fix annoyances

Without this patch there are a number of annoyances with the make
checks.  This patch fixes the following:

 1. Checks execute against .git/ and .terraform/ directories
 2. terraform validate executes redundantly

The patch refactors make.sh to use a general find | xargs pattern with a
single function to manage exclusion lists.

check_shebangs is removed because `#! /bin/bash -e` should not be used
because it does not take effect if the script is called via bash
<filename>.  Instead, calling `set -e` is preferred.

morgante commented on helpers/combine_docfiles.sh:
```diff
@@ -0,0 +1,133 @@
+#! /bin/bash
```
I think moving this into Bash from Python reduces ease of updates and shouldn't be done lightly.

jeffmccune replied:
Cool, I'll fix the python.  I'm sure there's a fix, just need to figure out a way to run the search and replace without deleting the trailing newline.

jeffmccune replied:
@morgante I've removed `combine_docfiles.sh` and improved the existing `combine_docfiles.py` to:

1. Use raw strings for regular expressions to avoid [W608](https://lintlyci.github.io/Flake8Rules/rules/W605.html) pyflake warnings.
2. Simple fix for newline: `output = ... + '\n'`

jeffmccune commented:
Closing this for the time being, the doc generation script needs to remain in Python.

morgante commented on Makefile:
```diff
@@ -54,10 +54,6 @@ check_docker:
 check_base_files:
 	@source test/make.sh && basefiles
 
-.PHONY: check_shebangs
```
Is there a reason we're dropping this check?

jeffmccune replied:
Yes, it's not run by default and it doesn't work correctly when run.  The check tries to check for `bash -e` in the `#!` line, but it doesn't.

If it did, this still isn't correct because `set -e` should be used instead.  The reason being, `bash foo.sh` will skip over the `#!` line and the `-e` flag.

morgante replied:
Got it.

jeffmccune replied:
For posterity, the dead code could be made to work with this change if desired:

```diff
diff --git a/test/make.sh b/test/make.sh
index e696ce8..e22a15e 100755
--- a/test/make.sh
+++ b/test/make.sh
@@ -18,7 +18,7 @@
 # shebang has a '- e' flag, which causes it
 # to exit on error
 function check_bash() {
-find . -name "*.sh" | while IFS= read -d '' -r file;
+find . -name "*.sh" | while IFS= read -r file;
 do
   if [[ "$file" != *"bash -e"* ]];
   then
```

The behavior is then:

```
make check_shebangs
./test/make.sh is missing shebang with -e
make: *** [check_shebangs] Error 1
```

morgante commented on test/make.sh:
```diff
@@ -14,63 +14,72 @@
-# This function checks to make sure that every
-# shebang has a '- e' flag, which causes it
-# to exit on error
-function check_bash() {
```
Why are we dropping this?

aaron-lane commented on test/make.sh:
```diff
@@ -14,63 +14,72 @@
-    exit 1;
-  fi;
-done;
+# find_files is a helper to exclude .git directories and match only regular
```
Should this also exclude `.terraform` directories?

jeffmccune replied:
Yes, it should.  Thanks for thinking of it, will add.

jeffmccune replied:
@aaron-lane Added the behavior of excluding `.terraform` directories.  Also modified the `make check_headers` task to use `find_files` because it was failing on files inside of .terraform directories for me.

aaron-lane commented on test/make.sh:
```diff
@@ -14,63 +14,72 @@
+# Compatibility with both GNU and BSD style xargs.
+compat_xargs() {
+  local compat=()
+  if xargs --no-run-if-empty </dev/null 2>/dev/null; then
```
```suggestion
  if xargs --help | grep --regexp '--no-run-if-empty' -; then
```

jeffmccune replied:
@aaron-lane I'm happy to make this change, but I'm getting the following undesirable behavior on my Mac with the change applied.  Could you help me understand the goal and I'll look into alternatives to achieve it?

```
Running shellcheck
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
Running flake8
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
Running go fmt and go vet
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
Running terraform validate
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
Running hadolint on Dockerfiles
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
Checking for required files LICENSE README.md
Testing the validity of the header check
..
----------------------------------------------------------------------
Ran 2 tests in 0.049s

OK
Checking file headers
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
The following lines have trailing whitespace
Generating markdown docs with terraform-docs
xargs: illegal option -- -
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
terraform-docs markdown .
terraform-docs markdown ./examples/simple_example
```

Compared to the original change:

```
Running shellcheck
Running flake8
Running go fmt and go vet
Running terraform validate
Running hadolint on Dockerfiles
Checking for required files LICENSE README.md
Testing the validity of the header check
..
----------------------------------------------------------------------
Ran 2 tests in 0.082s

OK
Checking file headers
The following lines have trailing whitespace
Generating markdown docs with terraform-docs
terraform-docs markdown .
terraform-docs markdown ./examples/simple_example
```

aaron-lane replied:
Apologies; context would have been helpful. I was trying to provide an alternative implementation with a more human readable approach. I guess the Mac version of xargs doesn't accept `-` as an argument, but I think the GNU version doesn't require it either so it can probably be safely dropped.

jeffmccune replied:
No problem!

I struggled to make this readable at first too, and try to avoid silently redirecting stderr if at all possible.

Would you approve something like this?

```bash
# Test if xargs is GNU or BSD style.  GNU xargs will succeed with status 0 when given
# --no-run-if-empty and no input on STDIN.  BSD xargs will fail and exit status non-zero
# If xargs fails, assume it is BSD style and proceed.  stderr is silently redirected to avoid
# console log spam.
if xargs --no-run-if-empty </dev/null 2>/dev/null; then
```

aaron-lane replied:
:+1:

aaron-lane reviewed (changes requested):
One question and one suggestion.

jeffmccune committed:
Use find_files to gain exclusion with check_headers

Without this patch the check_headers task descends into .terraform and
.git directories which causes false positive errors.

This patch switches check_headers to the find_files method, which has a
shared exclusion list containing .git and .terraform.

jeffmccune committed:
Fix combine_docfiles.py W605 and newline error

This patch fixes three issues with the make generate_docs task:

 1. Directories with *.tf files and no README.md throw errors.
 2. Newline at EOF is stripped from README.md
 3. W605 pylint warnings on combine_docfiles.py

jeffmccune committed:
Add .terraform to the make check exclusion list

aaron-lane commented on helpers/combine_docfiles.py:
```diff
@@ -50,6 +50,8 @@
 replace_content = groups[0] + groups[1]
 
 # Find where to put the replacement content, overwrite the input file
+# TODO(jmccune) fix AttributeError: 'NoneType' object has no attribute 'groups'
```
We should open issues rather than adding TODOs, if they aren't reasonable to address in the existing pull request.

jeffmccune replied:
Removed this TODO and opened a bug.

morgante commented on test/make.sh:
```diff
@@ -80,20 +107,27 @@ function check_trailing_whitespace() {
+    tmpfile="$(maketemp)"
+    echo "terraform-docs markdown ${path}"
+    terraform-docs markdown "${path}" > "${tmpfile}"
+    if ! [[ -e "${path}/README.md" ]]; then
```
I'm not sure this is the right approach - if there isn't already a README, we should just not generate docs instead of creating a README.

morgante replied:
@aaron-lane @adrienthebo What do you think?

jeffmccune replied:
I agree with you Morgante, we could just as easily skip over a directory with no readme.  I think that's best, with a clear output message to the console like `"Skipping ${path} because no README.md exists"`

The reason I took this approach of creating the readme is because that's what the current behavior is, so what you're seeing is an attempt to follow the spirit and intent.

morgante replied:
In that case, I'm happy to merge as is but please provide a follow-up PR which changes that behavior.

jeffmccune replied:
Done deal, behavior of skipping directories without README.md files is at: https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/6
